### PR TITLE
journalctl: fix output when --lines is used with --since or --until

### LINKED ISSUE #26657: journalctl -n misbehaves when combined with --since or --until
### systemd version the issue has been seen with

systemd-251.11-2.fc37.x86_64

### Used distribution

Fedora 37

### Linux kernel version used

6.1.9-200.fc37.x86_64

### CPU architectures issue was seen on

x86_64

### Component

journalctl

### Expected behaviour you didn't see

the man page of journalctl states that `-n` shows "the most recent journal events".

given there are more than 10 log lines in every time interval:
* when running `journalctl --since today -n 10`, I expect to see the same as `journalctl -n 10` 
* when running `journalctl --until today -n 10`, I expect to see the last 10 entries of yesterday (e.g. 23:59...).

### Unexpected behaviour you saw

* `journalctl --since today -n 10` returns the 10 oldest entries of today (same as `journalctl --since today | head -n 10`)
* `journalctl --until today -n 10` returns no entries at all

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by YHNdnzj**
The new logic is described in commit messages :)

Fixes #26657

### CODE REVIEW

File: src/journal/journalctl.c [hunk 1]
```diff
@@ -2571,14 +2577,26 @@ static int run(int argc, char *argv[]) {
...
 
                                 r = sd_journal_get_realtime_usec(j, &usec);
                                 if (r < 0)
                                         return log_error_errno(r, "Failed to determine timestamp: %m");
-                                if (usec < arg_since)
-                                        break;
+                                if (usec < arg_since) {
+                                        if (arg_reverse)
+                                                break; /* Reached the earliest entry */
+                                        else { /* arg_lines >= 0 (!since_seeked) */
```
Comment (yuwata) [inline src/journal/journalctl.c]: Unnecessary `else`

File: src/journal/journalctl.c [hunk 2]
```diff
@@ -2482,12 +2482,21 @@ static int run(int argc, char *argv[]) {
...
+                 * arg_lines and search afterwards from there. */
+
                 r = sd_journal_seek_realtime_usec(j, arg_until);
                 if (r < 0)
                         return log_error_errno(r, "Failed to seek to date: %m");
 
-                r = sd_journal_previous(j);
+                if (arg_reverse)
+                        r = sd_journal_previous(j);
+                else if (arg_lines >= 0)
```
Comment (yuwata) [inline src/journal/journalctl.c]: Unnecessary `if (arg_lines >= 0)`

File: src/journal/journalctl.c [hunk 3]
```diff
@@ -2552,7 +2561,7 @@ static int run(int argc, char *argv[]) {
                                         break;
                         }
 
-                        if (arg_until_set && !arg_reverse) {
+                        if (arg_until_set && !arg_reverse && arg_lines >= 0) {
```
Comment (yuwata) [inline src/journal/journalctl.c]: Is this change necessary?? It seems this change breaks `--until` _without_ `--lines`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/journal/journalctl.c (hunk 3 above)
Comment (YHNdnzj) [reply src/journal/journalctl.c]: Ouch, I got this inverted. It should be `&& arg_lines < 0` because otherwise the counter will get us covered.

**Review by yuwata — APPROVED** [review]
LGTM.

**Comment by YHNdnzj** [comment]
@yuwata Just made a minor change to the second commit, maybe have another look?

**Review by yuwata — APPROVED** [review]
Still LGTM.

**Pull request merged**